src/vs/workbench/parts/extensions/electron-browser/extensionsActions.ts

```diff
@@ -1709,6 +1726,64 @@ export class ConfigureRecommendedExtensionsCommandsContributor extends Disposabl
 			},
 			when: this.workspaceFolderContextKey
 		});
+
+		CommandsRegistry.registerCommand(AddToWorkspaceRecommendationsAction.ID, serviceAccesor => {
+			this.boundAddToRecommendationsContextKey.set(false);
+			const extensionId = (this.editorService.activeEditor as ExtensionsInput).extension.id;
+			serviceAccesor
+				.get(IInstantiationService)
+				.createInstance(
+					AddToWorkspaceRecommendationsAction,
+					AddToWorkspaceRecommendationsAction.ID,
+					AddToWorkspaceRecommendationsAction.LABEL
+				)
+				.run(extensionId)
+				.then(
+					(quickPickCancelled) => this.boundAddToRecommendationsContextKey.set(quickPickCancelled),
```

Comment:
I'd prefer just calling the update function in these cases, to keep the enablement logic a bit more self contained. It is a few file accesses, but this should be happening infrequently enough that that isn't a concern.

Reply:
What's controlling the visibility of these three commands is not in the instance of the action, it is the bounded `ContextKey` that's passed to the `when` property in `MenuRegistry.appendMenuItem`. Also, all of these actions are instantiated at the moment they are called every time and then immediately go out of scope, they don't stick around. There's nothing we could do in `update` to handle the enablement logic because of this. In fact, the previous two actions have to repeat the enablement logic here too, it's just not very noticeable because it's a one-liner for both of them (see https://github.com/reyronald/vscode/blob/1b8ff839fb1910d57c68a6516897a06a18aaa9dd/src/vs/workbench/parts/extensions/electron-browser/extensionsActions.ts#L1686 and https://github.com/reyronald/vscode/blob/1b8ff839fb1910d57c68a6516897a06a18aaa9dd/src/vs/workbench/parts/extensions/electron-browser/extensionsActions.ts#L1690)

The reason why the other two actions DO have an `update` and try to keep the `this.enabled` field updated is because they are being reused as UI buttons somewhere else (not just as commands). This is not the case with the new action and setting `this.enabled` to false has no effect on wether this will show up in the menu or not, that's why I had to move it up from the action itself to here.

Would you like us to put it inside again as public static methods that can be accessed from here or something similar?

Reply:
@reyronald 

The need to disable/hide this action if any of the workspace already has the current extension as recommendation arose when this was a button. As a button which clearly calls to action, we wanted to avoid it being shown unless absolutely required.

Now that this is in the command pallet and comes to the user's sight only when the user explicitly looks for it, I believe we can relax the conditions. If the action is indeed performed when the extension is already recommended, we can show a simple notification saying the same.

I'll push a commit simplifying this.